src/ppx/scene/scene_node.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/scene/scene_node.h"
+
+namespace ppx {
+namespace scene {
+
+// -------------------------------------------------------------------------------------------------
+// Node
+// -------------------------------------------------------------------------------------------------
+Node::Node(scene::Scene* pScene)
+    : mScene(pScene)
+{
+}
+
+Node::~Node()
+{
+}
+
+void Node::SetVisible(bool visible)
+{
+    mVisible = visible;
+}
+
+const float4x4& Node::GetEvaluatedMatrix() const
+{
+    if (mEvaluatedDirty) {
+        float4x4 parentEvaluatedMatrix = float4x4(1);
+        if (!IsNull(mParent)) {
+            parentEvaluatedMatrix = mParent->GetEvaluatedMatrix();
+        }
+        auto& concatenatedMatrix = GetConcatenatedMatrix();
+        mEvaluatedMatrix         = parentEvaluatedMatrix * concatenatedMatrix;
+        mEvaluatedDirty          = false;
+    }
+    return mEvaluatedMatrix;
+}
+
+void Node::SetParent(scene::Node* pNewParent)
+{
+    mParent = pNewParent;
```

Comment:
should this update mEvaluatedDirty?

Reply:
Thanks. One day I'll remember this function exists. Fixed.